File: opinionpulse/opinionpulse-backend/app/schemas/crisis.py
"""Pydantic schemas for Early-Warning Pulse & Crisis Radar."""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

Quadrant = Literal["quiet", "noise", "watch", "crisis"]


class CrisisNarrative(BaseModel):
    id: str
    label: str
    summary: str
    severity: Literal["low", "medium", "high", "critical"]
    negative_pct: float
    mention_count: int
    primary_platform: str
    example_snippet: str
    example_url: Optional[str] = None


class TimelineNode(BaseModel):
    id: str
    platform: str
    title: str
    snippet: str
    source_url: Optional[str] = None
    posted_at: Optional[str] = None
    minutes_after_origin: Optional[int] = None
    role: Literal["origin", "spread", "amplification"] = "spread"


class RadarPoint(BaseModel):
    watch_id: str
    keyword: str
    enabled: bool
    volume_score: float = Field(description="0–100: how many mentions in the last 30 minutes")
    velocity_score: float = Field(description="0–100: how fast negativity is accelerating")
    quadrant: Quadrant
    status_label: str
    status_explanation: str
    mention_count_30m: int
    negative_count_30m: int
    negative_pct_30m: float
    baseline_negative_30m: float
    last_scanned_at: Optional[str] = None
    in_crisis: bool = False


class CrisisRadarResponse(BaseModel):
    points: list[RadarPoint]
    legend: dict[str, str]
    last_updated: str
    scan_interval_minutes: int


class CrisisDetailResponse(BaseModel):
    watch_id: str
    keyword: str
    quadrant: Quadrant
    volume_score: float
    velocity_score: float
    status_label: str
    status_explanation: str
    mention_count_30m: int
    negative_count_30m: int
    negative_pct_30m: float
    narratives: list[CrisisNarrative]
    timeline: list[TimelineNode]
    recent_events: list["CrisisEventOut"]
    last_scanned_at: Optional[str] = None


class CrisisEventOut(BaseModel):
    id: str
    query: str
    quadrant: Quadrant
    volume_score: float
    velocity_score: float
    status_label: str
    summary: Optional[str] = None
    alert_sent: bool
    created_at: datetime


class CrisisScanResponse(BaseModel):
    watch_id: str
    keyword: str
    scanned: bool
    quadrant: Quadrant
    message: str
    detail: Optional[CrisisDetailResponse] = None

File: opinionpulse/opinionpulse-frontend/src/api/crisis.ts
import { apiRequest } from "@/api/client"

export type CrisisQuadrant = "quiet" | "noise" | "watch" | "crisis"

export type RadarPoint = {
  watch_id: string
  keyword: string
  name?: string | null
  enabled: boolean
  volume_score: number
  velocity_score: number
  quadrant: CrisisQuadrant
  status_label: string
  status_explanation: string
  mention_count_30m: number
  negative_count_30m: number
  negative_pct_30m: number
  baseline_negative_30m: number
  baseline_mentions_30m?: number
  negative_spike_multiplier?: number
  volume_spike_multiplier?: number
  spike_label?: string
  spike_severity?: string
  last_scanned_at: string | null
  in_crisis: boolean
}

export type CrisisNarrative = {
  id: string
  label: string
  summary: string
  severity: "low" | "medium" | "high" | "critical"
  negative_pct: number
  mention_count: number
  primary_platform: string
  example_snippet: string
  example_url?: string | null
}

export type TimelineNode = {
  id: string
  platform: string
  title: string
  snippet: string
  source_url?: string | null
  posted_at?: string | null
  minutes_after_origin?: number | null
  role: "origin" | "spread" | "amplification"
}

export type CrisisEvent = {
  id: string
  query: string
  quadrant: CrisisQuadrant
  volume_score: number
  velocity_score: number
  status_label: string
  summary?: string | null
  alert_sent: boolean
  created_at: string
}

export type CrisisRadarResponse = {
  points: RadarPoint[]
  legend: Record<string, string>
  last_updated: string
  scan_interval_minutes: number
}

export type CrisisDetailResponse = {
  watch_id: string
  keyword: string
  quadrant: CrisisQuadrant
  volume_score: number
  velocity_score: number
  status_label: string
  status_explanation: string
  mention_count_30m: number
  negative_count_30m: number
  negative_pct_30m: number
  narratives: CrisisNarrative[]
  timeline: TimelineNode[]
  recent_events: CrisisEvent[]
  last_scanned_at?: string | null
}

export async function getCrisisRadar(): Promise<CrisisRadarResponse> {
  return apiRequest<CrisisRadarResponse>("/api/crisis/radar", { auth: true })
}

export async function getCrisisDetail(watchId: string): Promise<CrisisDetailResponse> {
  return apiRequest<CrisisDetailResponse>(`/api/crisis/detail/${watchId}`, { auth: true })
}

export async function scanCrisisWatch(watchId: string): Promise<{
  watch_id: string
  keyword: string
  scanned: boolean
  quadrant: CrisisQuadrant
  message: string
  detail?: CrisisDetailResponse
}> {
  return apiRequest(`/api/crisis/scan/${watchId}`, { method: "POST", auth: true })
}

export async function listCrisisEvents(): Promise<CrisisEvent[]> {
  return apiRequest<CrisisEvent[]>("/api/crisis/events", { auth: true })
}
